Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -86,240 +106,1237 @@ local macro "pderiv_simp" : tactic =>
+        · simp only [add_of_Y_eq hP hPz hQz hxy.left hy hxy.right, nonsingular_smul,
+            nonsingular_zero]
+        · simp only [add_of_Y_ne hP.left hQ.left hPz hQz hxy.left hy, nonsingular_smul,
+            nonsingular_zero]
+
+variable (V) in
+/-- The addition of two point classes. If `P` is a point representative,
+then `W.addMap ⟦P⟧ ⟦Q⟧` is definitionally equivalent to `W.add P Q`. -/
+noncomputable def addMap (P Q : PointClass R) : PointClass R :=
+  Quotient.map₂ V.add (fun _ _ hP _ _ hQ => add_equiv hP hQ) P Q
+
+lemma addMap_eq (P Q : Fin 3 → R) : V.addMap ⟦P⟧ ⟦Q⟧ = ⟦V.add P Q⟧ :=
+  rfl
+
+lemma addMap_of_Z_eq_zero_left {P : Fin 3 → F} {Q : PointClass F} (hP : W.Nonsingular P)
+    (hQ : W.NonsingularLift Q) (hPz : P z = 0) : W.addMap ⟦P⟧ Q = Q := by
+  rcases Q with ⟨Q⟩
+  by_cases hQz : Q z = 0
+  · erw [addMap_eq, add_of_Z_eq_zero hP hQ hPz hQz, smul_eq, Quotient.eq]
+    exact Setoid.symm <| equiv_zero_of_Z_eq_zero hQ hQz
+  · erw [addMap_eq, add_of_Z_eq_zero_left hP hPz hQz, smul_eq]
+    rfl
+
+lemma addMap_of_Z_eq_zero_right {P : PointClass F} {Q : Fin 3 → F} (hP : W.NonsingularLift P)
+    (hQ : W.Nonsingular Q) (hQz : Q z = 0) : W.addMap P ⟦Q⟧ = P := by
+  rcases P with ⟨P⟩
+  by_cases hPz : P z = 0
+  · erw [addMap_eq, add_of_Z_eq_zero hP hQ hPz hQz, smul_eq, Quotient.eq]
+    exact Setoid.symm <| equiv_zero_of_Z_eq_zero hP hPz
+  · erw [addMap_eq, add_of_Z_eq_zero_right hQ hPz hQz, smul_eq]
+    rfl
+
+lemma addMap_of_Y_eq {P Q : Fin 3 → F} (hP : W.Nonsingular P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) (hx : P x * Q z ^ 2 = Q x * P z ^ 2)
+    (hy' : P y * Q z ^ 3 = W.negY Q * P z ^ 3) : W.addMap ⟦P⟧ ⟦Q⟧ = ⟦![1, 1, 0]⟧ := by
+  by_cases hy : P y * Q z ^ 3 = Q y * P z ^ 3
+  · rw [addMap_eq, add_of_Y_eq hP hPz hQz hx hy hy', smul_eq]
+  · rw [addMap_eq, add_of_Y_ne hP.left hQ hPz hQz hx hy, smul_eq]
+
+lemma addMap_of_Z_ne_zero {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) (hxy : P x * Q z ^ 2 = Q x * P z ^ 2 → P y * Q z ^ 3 ≠ W.negY Q * P z ^ 3) :
+    W.addMap ⟦P⟧ ⟦Q⟧ =
+      ⟦![W.toAffine.addX (P x / P z ^ 2) (Q x / Q z ^ 2)
+          (W.toAffine.slope (P x / P z ^ 2) (Q x / Q z ^ 2) (P y / P z ^ 3) (Q y / Q z ^ 3)),
+        W.toAffine.addY (P x / P z ^ 2) (Q x / Q z ^ 2) (P y / P z ^ 3)
+          (W.toAffine.slope (P x / P z ^ 2) (Q x / Q z ^ 2) (P y / P z ^ 3) (Q y / Q z ^ 3)),
+        1]⟧ := by
+  by_cases hx : P x * Q z ^ 2 = Q x * P z ^ 2
+  · rw [addMap_eq, add_of_Y_ne' hP hQ hPz hQz hx <| hxy hx, smul_eq]
+  · rw [addMap_eq, add_of_X_ne hP hQ hPz hQz hx, smul_eq]
+
+lemma nonsingularLift_addMap {P Q : PointClass F} (hP : W.NonsingularLift P)
+    (hQ : W.NonsingularLift Q) : W.NonsingularLift <| W.addMap P Q := by
+  rcases P, Q with ⟨⟨_⟩, ⟨_⟩⟩
+  exact nonsingular_add hP hQ
+
+end Addition
+
+section Point
+
+/-! ### Nonsingular rational points -/
+
+/-- A nonsingular rational point on `W`. -/
+structure Point where
```

Comment:
```suggestion
variable (V) in
/-- A nonsingular rational point on `V`. -/
@[ext] structure Point where
```
Please correct references to `W` in the docstrings that should actually be `V`.

Reply:
I'm actually meaning to ask for suggestions here. I think `V` is a horrible variable for a Weierstrass curve (over a ring), but I cannot think of another alphabet to use. Should I instead use `W'` over a ring and `W` over a field?

Reply:
`W'` would be fine, or you could introduce `(W : Jacobian R)` globally and `(W : Jacobian F)` only in sections that deal with fields, and it will shadow the global W within the section. However, beware that you might be unable to change explicitness of W within the sections, see [lean4#3227](https://github.com/leanprover/lean4/issues/3227).

Reply:
I'm aware of that bug. I made the distinction of `V`/`W` precisely for this reason.